src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/13.diff.sql

```diff
@@ -0,0 +1,424 @@
+    @resourceTypeId smallint,
+    @startResourceSurrogateId bigint,
+    @endResourceSurrogateId bigint,
+    @batchSize int
+AS
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DELETE Top(@batchSize) FROM dbo.Resource
+    WHERE ResourceTypeId = @resourceTypeId AND ResourceSurrogateId >= @startResourceSurrogateId AND ResourceSurrogateId < @endResourceSurrogateId
+
+    COMMIT TRANSACTION
+
+    return @@rowcount
+GO
+
+/*************************************************************
+    Stored procedures for batch delete ResourceWriteClaims
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     BatchDeleteResourceWriteClaims
+--
+-- DESCRIPTION
+--     Batch delete ResourceWriteClaims
+--
+-- PARAMETERS
+--     @startResourceSurrogateId
+--         * The start ResourceSurrogateId
+--     @endResourceSurrogateId
+--         * The end ResourceSurrogateId
+--     @batchSize
+--         * Max batch size for delete operation
+CREATE OR ALTER PROCEDURE dbo.BatchDeleteResourceWriteClaims
+    @startResourceSurrogateId bigint,
+    @endResourceSurrogateId bigint,
+    @batchSize int
+AS
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DELETE Top(@batchSize) FROM dbo.ResourceWriteClaim
+    WHERE ResourceSurrogateId >= @startResourceSurrogateId AND ResourceSurrogateId < @endResourceSurrogateId
+
+    COMMIT TRANSACTION
+
+    return @@rowcount
+GO
+
+
+/*************************************************************
+    Stored procedures for batch delete ResourceParams
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     BatchDeleteResourceParams
+--
+-- DESCRIPTION
+--     Batch delete ResourceParams
+--
+-- PARAMETERS
+--     @tableName
+--         * Resource params table name
+--     @resourceTypeId
+--         * Resource type id
+--     @startResourceSurrogateId
+--         * The start ResourceSurrogateId
+--     @endResourceSurrogateId
+--         * The end ResourceSurrogateId
+--     @batchSize
+--         * Max batch size for delete operation
+CREATE OR ALTER PROCEDURE dbo.BatchDeleteResourceParams
+    @tableName nvarchar(128),
+    @resourceTypeId smallint,
+    @startResourceSurrogateId bigint,
+    @endResourceSurrogateId bigint,
+    @batchSize int
+AS
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    DECLARE @Sql NVARCHAR(MAX);
+    DECLARE @ParmDefinition NVARCHAR(512);
+
+    SET @sql = N'DELETE TOP(@BatchSizeParam) FROM ' + @tableName	+ N' WHERE ResourceTypeId = @ResourceTypeIdParam AND ResourceSurrogateId >= @StartResourceSurrogateIdParam AND ResourceSurrogateId < @EndResourceSurrogateIdParam'
+    SET @parmDefinition = N'@BatchSizeParam int, @ResourceTypeIdParam smallint, @StartResourceSurrogateIdParam bigint, @EndResourceSurrogateIdParam bigint'; 
```

Comment:
This dynamic SQL is pretty risky -- is there an alternative that would avoid SQL injection risks?

Reply:
Table name and index name can not  be used as parameters in dynamic sql. For table name looks like we can use OBJECT_NAME(OBJECT_ID(@tableName)) to prevent SQL injection risks.